source/interface/azure_iot_mqtt.h

```diff
@@ -0,0 +1,380 @@
+     * @brief Message payload.
+     */
+    const void * pvPayload;
+
+    /**
+     * @brief Message payload length.
+     */
+    size_t xPayloadLength;
+} AzureIoTMQTTPublishInfo_t;
+
+typedef struct AzureIoTMQTTDeserializedInfo
+{
+    uint16_t usPacketIdentifier;                 /** @brief Packet ID of deserialized packet. */
+    AzureIoTMQTTPublishInfo_t * pxPublishInfo;   /** @brief Pointer to deserialized publish info. */
+    AzureIoTMQTTResult_t xDeserializationResult; /** @brief Return code of deserialization. */
+} AzureIoTMQTTDeserializedInfo_t;
+
+typedef struct AzureIoTMQTTPacketInfo
+{
+    /**
+     * @brief Type of incoming MQTT packet.
+     */
+    uint8_t ucType;
+
+    /**
+     * @brief Remaining serialized data in the MQTT packet.
+     */
+    uint8_t * pucRemainingData;
+
+    /**
+     * @brief Length of remaining serialized data.
+     */
+    size_t xRemainingLength;
+} AzureIoTMQTTPacketInfo_t;
+
+typedef AzureIoTMQTT_t * AzureIoTMQTTHandle_t;
+
+typedef uint32_t ( * AzureIoTMQTTGetCurrentTimeFunc_t )( void );
+
+typedef void ( * AzureIoTMQTTEventCallback_t )( AzureIoTMQTTHandle_t pContext,
+                                                struct AzureIoTMQTTPacketInfo * pxPacketInfo,
+                                                struct AzureIoTMQTTDeserializedInfo * pxDeserializedInfo );
+
+
+/**
+ * @brief Initialize an AzureIoTMQTT context.
+ *
+ * @param[in] xContext The context to initialize.
+ * @param[in] pxTransportInterface The transport interface to use with the context.
+ * @param[in] xGetTimeFunction The time utility function to use with the context.
+ * @param[in] xUserCallback The user callback to use with the context to
+ * notify about incoming packet events.
+ * @param[in] pucNetworkBuffer Network buffer provided for the context.
+ * @param[in] ulNetworkBufferLength Length of network buffer.
+ *
+ * @return An #AzureIoTMQTTResult_t with the result of the operation.
+ *      - eAzureIoTMQTTBadParameter if invalid parameters are passed.
+ *      - eAzureIoTMQTTSuccess otherwise.
+ */
+AzureIoTMQTTResult_t AzureIoTMQTT_Init( AzureIoTMQTTHandle_t xContext,
+                                        const AzureIoTTransportInterface_t * pxTransportInterface,
+                                        AzureIoTMQTTGetCurrentTimeFunc_t xGetTimeFunction,
+                                        AzureIoTMQTTEventCallback_t xUserCallback,
+                                        uint8_t * pucNetworkBuffer,
+                                        uint32_t ulNetworkBufferLength );
+
+
+/**
+ * @brief Establish an MQTT session.
+ *
+ * This function will send MQTT CONNECT packet and receive a CONNACK packet. The
+ * send and receive from the network is done through the transport interface.
+ *
+ * @param[in] xContext Initialized MQTT context.
+ * @param[in] pxConnectInfo MQTT CONNECT packet information.
+ * @param[in] pxWillInfo Last Will and Testament. Pass NULL if Last Will and
+ * Testament is not used.
```

Comment:
Nit - be more explicit about "is not used."  I tihnk what this means is that "Azure IoT Hub does not currently support Will messages" or something like that correct?